File: src/CSharp/101-200/131.PalindromePartitioning.cs

using System;
using System.Text;
using System.Collections.Generic;


namespace PeefyLeetCode.PalindromePartitioning
{
    public class Solution
    {
        List<IList<string>> list = new List<IList<string>>();
        string s;
        public void dfs(List<string> ll, int index)
        {
            if (index == s.Length)
            {
                list.Add(new List<string>(ll));
                return;
            }
            for (int i = index; i < s.Length; i++)
            {
                if (isPalindrome(index, i))
                {
                    ll.Add(s.Substring(index, i + 1 - index));
                    dfs(ll, i + 1);
                    ll.RemoveAt(ll.Count - 1);
                }

            }
        }

        public bool isPalindrome(int start, int end)
        {
            while (start < end)
            {
                if (s[start] != s[end])
                    return false;
                start++;
                end--;
            }
            return true;
        }
        public IList<IList<string>> Partition(string s)
        {
            this.s = s;
            List<string> ll = new List<string>();
            dfs(ll, 0);
            return list;
        }
    }
}

File: src/Java/101-200/131.PalindromePartitioning.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

class Solution {
    List<List<String>> list = new ArrayList<>();
    String s;

    public List<List<String>> partition(String s) {
        this.s = s;
        List<String> ll = new ArrayList<>();
        dfs(ll, 0);
        return list;
    }

    public void dfs(List<String> ll, int index) {
        if (index == s.length()) {
            list.add(new ArrayList<>(ll));
            return;
        }
        for (int i = index; i < s.length(); i++) {
            if (isPalindrome(index, i)) {
                ll.add(s.substring(index, i + 1));
                dfs(ll, i + 1);
                ll.remove(ll.size() - 1);
            }

        }
    }

    public boolean isPalindrome(int start, int end) {
        while (start < end) {
            if (s.charAt(start) != s.charAt(end))
                return false;
            start++;
            end--;
        }
        return true;
    }
}
